runtime/src/main/java/org/corfudb/runtime/clients/LayoutHandler.java

```diff
@@ -94,4 +105,116 @@ private static Object handleAlreadyBootstrap(CorfuMsg msg,
+     */
+    @ResponseHandler(type = PayloadCase.PREPARE_LAYOUT_RESPONSE)
+    private static Object handlePrepareLayoutResponse(ResponseMsg msg, ChannelHandlerContext ctx,
+                                                      IClientRouter r) {
+        PrepareLayoutResponseMsg prepareLayoutMsg = msg.getPayload().getPrepareLayoutResponse();
+        PrepareLayoutResponseMsg.Type type = prepareLayoutMsg.getRespType();
+        long rank = prepareLayoutMsg.getRank();
+        Layout layout = CorfuProtocolCommon.getLayout(prepareLayoutMsg.getLayout());
+
+        switch (type) {
+            case ACK: return new LayoutPrepareResponse(rank, layout);
+            case REJECT: throw new OutrankedException(rank, layout);
+                // TODO for INVALID
+            default: throw new UnsupportedOperationException("Response handler not provided");
+        }
+    }
+
+    /**
+     * Handle a propose layout response from the server.
+     *
+     * @param msg The propose layout response message.
+     * @param ctx The context the message was sent under.
+     * @param r A reference to the router.
+     * @return True if ACK, throw an {@link OutrankedException} if REJECT.
+     */
+    @ResponseHandler(type = PayloadCase.PROPOSE_LAYOUT_RESPONSE)
+    private static Object handleProposeLayoutResponse(ResponseMsg msg, ChannelHandlerContext ctx,
+                                                      IClientRouter r) {
+        ProposeLayoutResponseMsg proposeLayoutMsg = msg.getPayload().getProposeLayoutResponse();
+        ProposeLayoutResponseMsg.Type type = proposeLayoutMsg.getRespType();
+        long rank = proposeLayoutMsg.getRank();
+
+        switch (type) {
+            case ACK: return true;
+            case REJECT: throw new OutrankedException(rank);
+                // TODO for INVALID
+            default: throw new UnsupportedOperationException("Response handler not provided");
+        }
+    }
+
+    /**
+     * Handle a propose layout response from the server.
+     *
+     * @param msg The propose layout response message.
+     * @param ctx The context the message was sent under.
+     * @param r A reference to the router.
+     * @return True if ACK, throw an {@link OutrankedException} if REJECT.
+     */
+    @ResponseHandler(type = PayloadCase.COMMIT_LAYOUT_RESPONSE)
+    private static Object handleCommitLayoutResponse(ResponseMsg msg, ChannelHandlerContext ctx,
+                                                     IClientRouter r) {
+        CommitLayoutResponseMsg commitLayoutMsg = msg.getPayload().getCommitLayoutResponse();
+        CommitLayoutResponseMsg.Type type = commitLayoutMsg.getRespType();
+
+        switch (type) {
+            case ACK: return true;
+            case NACK: return false;
+            // TODO for INVALID
+            default: throw new UnsupportedOperationException("Response handler not provided");
+        }
```

Comment:
Why don't you just use a boolean to represent this.

```proto
message CommitLayoutResponseMsg {
      bool committed = 1;
}

```

Reply:
I think when @zfrenette design these message types, enum types were used to make it easier for extension. For example when the message type needs to change sometime in the future, to model something like ACK, NACK-FAILURE-REASON-A, NACK-FAILURE-REASON-B, enum might provide more flexibility and easier for backward compatibility. But definitely using a boolean would make it cleaner at this time. We might need to make a tradeoff here.

Reply:
I would just go with a boolean, what do you think @zfrenette